zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -0,0 +1,1107 @@
+    mut entries: impl Iterator<Item = ScanRange>,
+) -> Result<(), SqliteClientError> {
+    let (to_create, to_delete_ends) = {
+        let mut suggested_stmt = conn.prepare_cached(
+            "SELECT block_range_start, block_range_end, priority
+            FROM scan_queue
+            WHERE (
+                -- the start is contained within the range
+                :start >= block_range_start
+                AND :start < block_range_end
+            )
+            OR (
+                -- the end is contained within the range
+                :end > block_range_start
+                AND :end <= block_range_end
+            )
+            OR (
+                -- start..end contains the entire range
+                block_range_start >= :start
+                AND block_range_end <= :end
+            )
+            ORDER BY block_range_end",
+        )?;
+
+        let mut rows = suggested_stmt.query(named_params![
+            ":start": u32::from(query_range.start),
+            ":end": u32::from(query_range.end),
+        ])?;
+
+        // Iterate over the ranges in the scan queue that overlaps the range that we have
+        // identified as needing to be fully scanned. For each such range add it to the
+        // spanning tree (these should all be nonoverlapping ranges, but we might coalesce
+        // some in the process).
+        let mut existing_ranges: Option<SpanningTree> = None;
+        let mut to_delete_ends: Vec<Value> = vec![];
+        while let Some(row) = rows.next()? {
+            let entry = ScanRange::from_parts(
+                Range {
+                    start: BlockHeight::from(row.get::<_, u32>(0)?),
+                    end: BlockHeight::from(row.get::<_, u32>(1)?),
+                },
+                {
+                    let code = row.get::<_, i64>(2)?;
+                    parse_priority_code(code).ok_or_else(|| {
+                        SqliteClientError::CorruptedData(format!(
+                            "scan priority not recognized: {}",
+                            code
+                        ))
+                    })?
+                },
+            );
+            to_delete_ends.push(Value::from(u32::from(entry.block_range().end)));
+            existing_ranges = if let Some(cur) = existing_ranges {
+                Some(cur.insert(entry))
+            } else {
+                Some(SpanningTree::Leaf(entry))
+            };
+        }
+
+        // Update the tree that we read from the database, or if we didn't find any ranges
+        // start with the scanned range.
+        let mut to_create = match (existing_ranges, entries.next()) {
+            (Some(cur), Some(entry)) => Some(cur.insert(entry)),
+            (None, Some(entry)) => Some(SpanningTree::Leaf(entry)),
+            (Some(cur), None) => Some(cur),
+            (None, None) => None,
+        };
+
+        for entry in entries {
+            to_create = to_create.map(|cur| cur.insert(entry));
+        }
```

Comment:
This can be simplified using the same technique as in the earlier row loop:
```suggestion
        for entry in entries {
            to_create = if let Some(cur) = to_create {
                Some(cur.insert(entry))
            } else {
                Some(SpanningTree::Leaf(entry))
            };
        }
```